### PR TITLE
[wasm][sdk] Threading support for the SDK

### PR DESCRIPTION
**Pull request by kjpou1**
**Addresses the loading of the threads runtime if the browser supports threads and the non threads runtime when it is not supported in the browser.**

**Add `mono_wasm_has_threading_support` function that can be called from JavaScript code.**
---
   - Tests whether the currently executing code was compiled with pthreads support enabled. If this function returns true, then the currently executing code was compiled with -s USE_PTHREADS=1 (and the current browser supports multithreading).

   example:

   ```
	console.log("threading support: " + MONO.mono_wasm_has_threading_support());
   ```

**[sdk] Add new property `IncludeMonoWasmThreads`**
---
   - Determines whether to include support for threads or not.
      - "None" - Do not include runtime for threads. - (Default)
      - "ThreadsOnly" - Only include runtime for threads support.
      - "All" - Include runtimes for both thread and non thread support.

**[sdk] Add new property `EnableAutoRuntimeLoading`**
---
   - Whether to enable support for automatically loading the correct wasm (threaded or non-threaded) runtime.
      - "True" - The `runtime.js` will attempt to load the correct runtime. 
      - "False" - The `runtime.js` will NOT attempt to load the correct runtime. 

   - `runtime.js` file now supports loading the correct `mono.js` when `<EnableAutoRuntimeLoading>` is `True`.  If threads are supported it will load the threads `mono.js` module instead of the non threads runtime modules.
      - If both runtimes are included then a new directory `threads` will be created to load the threads runtime.

- To load the correct runtime if not using the `EnableAutoRuntimeLoading` the following can be used as an example.

```
    <script>
        var firstScript = document.getElementsByTagName('script')[0],
            js = document.createElement('script');
        js.src = "mono.js";
        if (typeof SharedArrayBuffer !== 'undefined')
            js.src = "threads/mono.js";
        firstScript.parentNode.insertBefore(js, firstScript);
    </script>

```

   - **Note: This is a breaking change.  Since `runtime.js` now takes care of loading the correct runtime modules,when `<EnableAutoRuntimeLoading>` is `True`, referencing `<script defer src="mono.js"></script>` from the html will now throw error.**

**Modify the template to add `<EnableMonoWasmThreads>None</EnableMonoWasmThreads>` by default.**
---

**Modify the browser tests to remove the `<script defer src="mono.js"></script>` references.**
---
   - This is now taken care of in the `runtime.js` generated by the SDK.

**Add two new docs.**
---
   - CHANGES.md
   - BREAKING-CHANGES.md

### CODE REVIEW

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@kjpou1 please rebase

**Comment by kjpou1** [comment]
Will do

File: sdks/wasm/src/library_mono.js [hunk 1]
```diff
@@ -376,7 +383,7 @@ var MonoSupportLib = {
 	mono_wasm_fire_bp: function () {
 		console.log ("mono_wasm_fire_bp");
 		debugger;
-	}
+	}
```
Comment (MaximLipnin) [inline sdks/wasm/src/library_mono.js]: nit: tab

File: sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt [hunk 2]
```diff
@@ -14,10 +29,19 @@ var Module = {
 	},
 };
 
-
+(function() {
```
Comment (MaximLipnin) [inline sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: nit: space before parenths (and a bit below)

File: sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt [hunk 3]
```diff
@@ -1,6 +1,21 @@
...
+	threadsRuntimeAvailable: <#=ThreadsRuntimeAvailable??false#>, 
+	locateMonoThreads: function (path, pref) {
+		return pref + path;
+	},
+	locateFile: function (path, pref) {
+		if (this.threadsRuntimeAvailable) {
+			if (typeof SharedArrayBuffer !== 'undefined') {
+				if (path.startsWith ("mono.js")) return this.locateMonoThreads (path, pref);
+				if (path.startsWith ("mono.wasm")) return this.locateMonoThreads (path, pref);
+				if (path === ("mono.worker.js")) return this.locateMonoThreads (path, pref);
```
Comment (MaximLipnin) [inline sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: nit: redundant parenths

File: sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt (hunk 2 above)
Comment (kjpou1) [reply sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: Maxim not sure I see this one.

Comment (MaximLipnin) [reply sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: need to be added :)

File: sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt (hunk 3 above)
Comment (kjpou1) [reply sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: not sure I see this

Comment (kjpou1) [reply sdks/wasm/sdk/Mono.WebAssembly.Build/RuntimeJs.tt]: got it

**Comment by lewing** [comment]
The threaded runtime isn't functional right now so this is on hold

**Comment by lewing** [comment]
closing because of conflicts, please rebase and reopen when ready

**Pull request closed without merging**